# ------------------------------------------------------------------------------
# AnnotateGenome
# 2020/10/02
# Ryan D. Crawford
# ------------------------------------------------------------------------------
#' Create a gff file for a bacterial genome
#' @description
#'   This function uses the RAST command line interface to annotate the 
#'   genome and generate a GFF-3 file. First, a genome type object is
#'   generated with RAST. Then the genome is processed, and the 
#'   features are exported to a GFF file. 
#' @param faPath Path to the fasta file containing the whole genome sequence
#'   to be annotated.
#' @param outDir Directory where the the gff and gto are to be written.
#' @param genomeName Optional name for the gff file. By default, the extension
#'   is removed from the fasta file and the remainder is used.
#' @param keepGto Optional bool to keep the "genome type object" created
#'   by RAST for further processing (see the rast documentation). False by 
#'   default
#' @return The path to the gff file
#' @export
#  -----------------------------------------------------------------------------

AnnotateGenome = function( faPath, outDir, genomeName, keepGto )
{
  # If no genome name was input, create one
  if ( missing( genomeName ) ) genomeName = ExtractGenomeNameFromPath( faPath )
  
  # If no output directory was passed as argument write to the 
  # current working directory
  # If now output directory was specified, write to the working directory
  if ( missing(outDir) )
  {
    outDir = ''
  # Make sure the output directory ends in a '/'
  } else if ( !grepl("/$", outDir) && outDir != '' ) {
    outDir = paste0( outDir, '/' )
  }
  
  # By default do not keep the genome type object generated by RAST
  if ( missing( keepGto ) ) keepGto = FALSE
  
  # Make the path to the genome type object to be created
  gtoPath = paste0( outDir, genomeName, ".gto" )
  
  # Create a genome type object
  createGenomeCmd = paste(
    "rast-create-genome",
    "--scientific-name", genomeName,
    "--genetic-code 11",
    "--domain Bacteria",
    "--contigs", faPath,
    '>', gtoPath
    )
  system( createGenomeCmd )
  
  # Process the genome 
  proGto = paste0( outDir, genomeName, "_pro.gto" )
  system( paste( "rast-process-genome <", gtoPath, '>', proGto ) )
  
  # Create the path to the ouptput gff file to be created
  gffPath = paste0( outDir, genomeName, ".gff" )
  
  # Export the genome features in gff-3 format
  system( paste( "rast-export-genome", "gff", '<', proGto, '>', gffPath ) )
  
  # If not requested to keep the temp files, delete the gto
  if ( !keepGto ) system( paste( "rm", gtoPath, proGto ) )
  
  return( gffPath )
}

# ------------------------------------------------------------------------------
