# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

CalcAlgnSubMatrix <- function(seqs) {
    .Call(`_cognac_CalcAlgnSubMatrix`, seqs)
}

CalcAlgnPartitionDists <- function(msaPath, method, genePartitions) {
    .Call(`_cognac_CalcAlgnPartitionDists`, msaPath, method, genePartitions)
}

ConcatenateAlignments <- function(concatAlgn, algn) {
    invisible(.Call(`_cognac_ConcatenateAlignments`, concatAlgn, algn))
}

#' @name CreateAlgnDistMat
#' @title Create Algnment Distance Matrix
#' @description
#'   This function reads in the path to a multiple sequence alignment. The
#'   pairwise alignment distances for each sequence are returned. The
#'   pairwise distances are calculated in parallel via the RcppParallel
#'   package.
#' @param msaPath Path to the alignment
#' @param method Method for calculating distance: "raw", or "shared."
#' @param isCore Logical to specify whether to remove gap positions from the
#'   alignment to create the core genome.
#' @return A numeric matrix
#' @export
NULL

CreateAlgnDistMat <- function(msaPath, method) {
    .Call(`_cognac_CreateAlgnDistMat`, msaPath, method)
}

CreateCognacRunData <- function(geneEnv, gfPaths, faPaths, faaPath) {
    invisible(.Call(`_cognac_CreateCognacRunData`, geneEnv, gfPaths, faPaths, faaPath))
}

CreateCoreGenomeDistMat <- function(msaPath) {
    .Call(`_cognac_CreateCoreGenomeDistMat`, msaPath)
}

DeletePartitions <- function(msaPath, delStart, delEnd, outPath) {
    invisible(.Call(`_cognac_DeletePartitions`, msaPath, delStart, delEnd, outPath))
}

ExtractGenomeNameFromPath <- function(path) {
    .Call(`_cognac_ExtractGenomeNameFromPath`, path)
}

GetGenomeNameWithExt <- function(path, ext) {
    .Call(`_cognac_GetGenomeNameWithExt`, path, ext)
}

#' @name FilterAlignmentPositions
#' @title Filter Alignment Positions
#' @description
#'   This function removes gaps and/or positions without sufficient varition
#'   from the alingment.
#' @param msaPath Path to the alignment
#' @param filterMsaPath Path to write the filtered alignment
#' @param minGapFrac Double representing the minimium fraction of gaps to
#'   remain in the alignment.Defaults to 0.01.
#' @param minSubThresh Integer representing the minimum number of
#'   substitutions to remain in the alignment. This is the number of instances
#'   of any minor allele to remain. Defaults to 0.
#' @return void
#' @export
NULL

FilterAlgnPositions <- function(msaPath, filterMsaPath, minGapFrac = 0.01, minSubThresh = 0L) {
    invisible(.Call(`_cognac_FilterAlgnPositions`, msaPath, filterMsaPath, minGapFrac, minSubThresh))
}

#' @name FilterPartitionedAlgnPositions
#' @title Filter Partitioned Algn Positions
#' @description
#'   This function
#' @param msaPath Path to the alignment
#' @param filterMsaPath Path to write the filtered alignment
#' @param genePositions Vector of gene partitions in the alignment
#' @param minGapFrac Double representing the minimium fraction of gaps to
#'   remain in the alignment.Defaults to 0.01.
#' @param minSubThresh Integer representing the minimum number of
#'   substitutions to remain in the alignment. This is the number of instances
#'   of any minor allele to remain. Defaults to 0.
#' @return Updated vector with the partions in the filtered alignment
#' @export
NULL

FilterPartitionedAlgnPositions <- function(msaPath, filterMsaPath, genePositions, minGapFrac = 0.01, minSubThresh = 0L) {
    .Call(`_cognac_FilterPartitionedAlgnPositions`, msaPath, filterMsaPath, genePositions, minGapFrac, minSubThresh)
}

FindIdenticalGenes <- function(genes, geneIds) {
    .Call(`_cognac_FindIdenticalGenes`, genes, geneIds)
}

GetGenomeId <- function(inStr) {
    .Call(`_cognac_GetGenomeId`, inStr)
}

ParseCdHit <- function(cdHitClstFile, isBinary, clSizeThesh, geneEnv) {
    invisible(.Call(`_cognac_ParseCdHit`, cdHitClstFile, isBinary, clSizeThesh, geneEnv))
}

ParseFasta <- function(faPath) {
    .Call(`_cognac_ParseFasta`, faPath)
}

TranslateAaAlgnToDna <- function(gffData, faPath, genePositions, genomeName, aaAlgn, outputFile) {
    invisible(.Call(`_cognac_TranslateAaAlgnToDna`, gffData, faPath, genePositions, genomeName, aaAlgn, outputFile))
}

# Register entry points for exported C++ functions
methods::setLoadAction(function(ns) {
    .Call('_cognac_RcppExport_registerCCallable', PACKAGE = 'cognac')
})
