# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' @name CreateAlgnDistMat
#' @title  Create Algnment Distance Matrix
#' @description
#'   This function reads in the path to a multiple sequence alignment. The
#'   pairwise alignment distances for each sequence are returned. The 
#'   pairwise distances are calculated in parallel via the RcppParallel 
#'   package.
#' @param Path to the alignment
#' @param Method for calculating distance: "raw", or "shared."
#' @param Logical to specify whether to remove gap positions from the alignment
#'   to create the core genome.
#' @return A numeric matrix
#' @export
NULL

CreateAlgnDistMat <- function(msaPath, method, isCore) {
    .Call(`_cognac_CreateAlgnDistMat`, msaPath, method, isCore)
}

CreateCognacRunData <- function(geneEnv, gfPaths, faPaths, faaPath) {
    invisible(.Call(`_cognac_CreateCognacRunData`, geneEnv, gfPaths, faPaths, faaPath))
}

ExtractGenomeNameFromPath <- function(path) {
    .Call(`_cognac_ExtractGenomeNameFromPath`, path)
}

GetGenomeNameWithExt <- function(path, ext) {
    .Call(`_cognac_GetGenomeNameWithExt`, path, ext)
}

FindIdenticalGenes <- function(genes, geneIds) {
    .Call(`_cognac_FindIdenticalGenes`, genes, geneIds)
}

GetGenomeId <- function(inStr) {
    .Call(`_cognac_GetGenomeId`, inStr)
}

ParseCdHit <- function(cdHitClstFile, isBinary, clSizeThesh, geneEnv) {
    invisible(.Call(`_cognac_ParseCdHit`, cdHitClstFile, isBinary, clSizeThesh, geneEnv))
}

ParseFasta <- function(faPath) {
    .Call(`_cognac_ParseFasta`, faPath)
}

TranslateAaAlgnToDna <- function(gffData, faPath, genePositions, genomeName, aaAlgn, outputFile) {
    invisible(.Call(`_cognac_TranslateAaAlgnToDna`, gffData, faPath, genePositions, genomeName, aaAlgn, outputFile))
}

# Register entry points for exported C++ functions
methods::setLoadAction(function(ns) {
    .Call('_cognac_RcppExport_registerCCallable', PACKAGE = 'cognac')
})
